File: QuasselDroidNG/src/main/java/de/kuschku/libquassel/primitives/serializers/StringListSerializer.java
package de.kuschku.libquassel.primitives.serializers;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.ByteChannel;
import java.util.ArrayList;
import java.util.List;

public class StringListSerializer implements PrimitiveSerializer<List<String>> {
    private static final StringListSerializer serializer = new StringListSerializer();

    private StringListSerializer() {
    }

    public static StringListSerializer get() {
        return serializer;
    }

    @Override
    public void serialize(final ByteChannel channel, final List<String> data) throws IOException {
        IntSerializer.get().serialize(channel, data.size());
        for (String element : data) {
            StringSerializer.get().serialize(channel, element);
        }
    }

    @Override
    public List<String> deserialize(final ByteBuffer buffer) throws IOException {
        final int size = IntSerializer.get().deserialize(buffer);
        final List<String> list = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            list.add(StringSerializer.get().deserialize(buffer));
        }
        return list;
    }
}

File: app/src/main/java/de/kuschku/quasseldroid_ng/protocol/primitive/serializer/StringListSerializer.kt
package de.kuschku.quasseldroid_ng.protocol.primitive.serializer

import de.kuschku.quasseldroid_ng.protocol.QStringList
import de.kuschku.quasseldroid_ng.protocol.Quassel_Features
import de.kuschku.quasseldroid_ng.util.nio.ChainedByteBuffer
import java.nio.ByteBuffer

object StringListSerializer : Serializer<QStringList?> {
  override fun serialize(buffer: ChainedByteBuffer, data: QStringList?,
                         features: Quassel_Features) {
    IntSerializer.serialize(buffer, data?.size ?: 0, features)
    data?.forEach {
      StringSerializer.UTF16.serialize(buffer, it, features)
    }
  }

  override fun deserialize(buffer: ByteBuffer, features: Quassel_Features): QStringList {
    return (0 until IntSerializer.deserialize(buffer, features)).map {
      StringSerializer.UTF16.deserialize(buffer, features)
    }.toMutableList()
  }

}
